src/coreclr/jit/layout.cpp

```diff
@@ -939,18 +1085,53 @@ void ClassLayoutBuilder::CopyInfoFrom(unsigned offset, ClassLayout* layout, bool
             SetGCPtr(startSlot + slot, layout->GetGCPtr(slot));
         }
     }
+}
 
-    if (copyPadding)
-    {
-        AddPadding(SegmentList::Segment(offset, offset + layout->GetSize()));
+//------------------------------------------------------------------------
+// CopyInfoGCFromMakeByref: Copy GC pointers from another layout,and change
+//   all gc references to be TYP_BYREF (TYPE_GC_BYREF)
+//
+// Arguments:
+//   offset      - Offset in this builder to start copy information into.
+//   layout      - Layout to get information from.
+//
+void ClassLayoutBuilder::CopyGCInfoFromMakeByref(unsigned offset, ClassLayout* layout)
+{
+    assert(offset + layout->GetSize() <= m_size);
 
-        for (const SegmentList::Segment& nonPadding : layout->GetNonPadding(m_compiler))
+    if (layout->GetGCPtrCount() > 0)
+    {
+        assert(offset % TARGET_POINTER_SIZE == 0);
+        unsigned startSlot = offset / TARGET_POINTER_SIZE;
+        for (unsigned slot = 0; slot < layout->GetSlotCount(); slot++)
         {
-            RemovePadding(SegmentList::Segment(offset + nonPadding.Start, offset + nonPadding.End));
+            CorInfoGCType gcType = layout->GetGCPtr(slot);
+            if (gcType == TYPE_GC_REF)
+            {
+                gcType = TYPE_GC_BYREF;
+            }
+            SetGCPtr(startSlot + slot, gcType);
         }
     }
 }
```

Comment:
Same comment as previously.. This feels awfully specific to escape analysis to add in the generic builder type.